### ISSUE TITLE
TRT INT8 Compilation Have Different Outputs Comparing to Modelopt Fake Q/DQ nodes for Resnet18

### ISSUE THREAD

**Issue by YixuanSeanZhou** [NVIDIA/TensorRT-Model-Optimizer#48]
When performing Resnet18 PTQ using TRT-modelopt, I encountered the following issue when compiling the model with TRT. 

First off, I started with a pretrained resnet18 from torchvision. I replaced the last fully connected layer to fit on my dataset (for example, CIFRA-10). I also updated all the skip layers (the plus) with a ElementwiseAdd layer and I defined its quantization layer as follow myself (code attached at the end). The reason I do this is to facilitate the Q/DQ fusion so that every layer can be in INT8. 

Then, when compiling the exported onnx model with TRT, I found that TRT outputs is very different from the fake Q/DQ model in python, and the fake Q/DQ onnx model as well when running with onnx runtime. (`np.allclose` with `1e-3` as the threshold failed) 

Visualizing the TRT engine, I think it is completely within my expectation with everything being fused as INT8 kernels. 
![trt_engine_0](https://github.com/user-attachments/assets/2a0f333d-02c1-4eb0-95f1-2e7a723d1b33)

I am unsure what's wrong and how to debug this issue. Could you please take a look? You can download the onnx model and the TRT engine here: 

https://file.io/GnuiEMNeebQ1

Thanks! 

Also attaching the implementation of a quantized skip connection
```
# Quantized Skip Connection

class QuantElementWiseAdd(ElementWiseAdd):
    def __init__(self, normalized_shape):
        super().__init__(normalized_shape)
        self._setup()

    def _setup(self):
        # Method to setup the quantizers
        self.input1_quantizer = TensorQuantizer()
        self.input2_quantizer = TensorQuantizer()
        self.output_quantizer = TensorQuantizer()

    def forward(self, input1, input2):
        # You can customize the quantizer placement anywhere in the forward method
        input1 = self.input1_quantizer(input1)
        input2 = self.input2_quantizer(input2)
        return self.output_quantizer(input1 + input2)
```

**Comment by i-riyad** [comment]
- From the TensorRT graph it looks okay. What is the top1 or top5 accuracy?
- Have you tried `modelopt.onnx.quantization` instead? Export the original pytorch model to ONNX and then quantize it like [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/onnx_ptq#quantize-an-onnx-model).

**Comment by YixuanSeanZhou** [comment]
Thanks for the response @riyadshairi979 So sorry for my delayed response, somehow i missed it in the inbox

top1 accuracy dropped by 2.3%, top 5 accuracy is no drop, but i don't think top 5 is good because i am doing CiFAR-10 with only 10 classes. 
`modelopt.onnx.quantization` unfortunately, this is not something we would like to consider. We are somewhat emphasizing in quantizing purely in Pytorch, for easier evaluation loop (we want to eval the quantization result in Pytorch, which is also why we were hoping for no gap between Pytorch and TRT). 

> top1 accuracy dropped by 2.3%,
This IMO is a somewhat significant drop. It means I will not be able to have confidence to claim parity between TRT model and Pytorch model. Comparing this with FP32 (no quantization), Pytorch and TRT differ by 0.1%, which is much more acceptable. 

What are other options we have to resolve this gap? 

Thanks in advance, really looking forward to your guidance!

**Comment by YixuanSeanZhou** [comment]
@riyadshairi979 

A bit more information: when bisecting the graph and try to generate layerwise diff, I noticed that the output seems to start diverge when it comes to the first conv in the first basic block. 

In TRT, the first Conv -> BN -> Maxpool is almost elementwise the same as this network in modelopt (torch). However, the error seems to happen after Conv after the Maxpool layer. 

Also just FYI: My TRT version is 8.6.1. I think it is recent enough. (We are not looking to upgrade to TRT 10 just yet)

**Comment by YixuanSeanZhou** [comment]
Also just FYI: My TRT version is 8.6.1. I think it is recent enough. (We are not looking to upgrade to TRT 10 just yet)

**Comment by i-riyad** [comment]
> when it comes to the first conv in the first basic block

Can you try excluding this layer from quantization? @realAsma does modelopt provides a way to configure exclusion of certain linear/conv layers?
Another way is to do ONNX graphsurgeon on the exported ONNX model to remove the QDQ nodes before that first Conv node.

**Comment by YixuanSeanZhou** [comment]
Hi @riyadshairi979, thank you so much for your reply. 

Yes, i think I can exclude this layer in modelopt through the qconfig. I will try that and let you know how it goes. However, my fear is other layers will have the same behaviors. 

Even if disabling quantization on this layer worked, this won't be scalable right? We are piloting with resnet18 as a toy model, which is why we can force dumping each layer's output in TRT, we won't be able to do it in a real model. 

I thought it should be expected that Pytorch modelopt output is very close to TRT output, is this assumption invalid? 

By any chance it could be a TRT issue on fusing the QDQ nodes into a conv block? Does it has anything to do with fusing quantized Conv -> BN -> Relu as described i[n here](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#qdq-placement-recs)? I wonder if this bug can be fixed on the TRT side? 

Thanks again!

**Comment by i-riyad** [comment]
> Then, when compiling the exported onnx model with TRT, I found that TRT outputs is very different from the fake Q/DQ model in python, and the fake Q/DQ onnx model as well when running with onnx runtime. 

Though in TensorRT there are `Conv -> BN -> Relu` etc. fusions and fake QDQ in torch does not simulate all those TRT optimizations, I agree 2.3% accuracy degradation is not expected. This is a very common fused layer in many CNN models and we haven't observed this much regression. May be this particular fused `Conv -> BN -> Relu` TRT kernel in int8 precision introduces quantization errors and most likely it is a bug which you should file [here](https://github.com/NVIDIA/TensorRT/issues).

**Comment by realAsma** [comment]
@YixuanSeanZhou 

> Then, when compiling the exported onnx model with TRT, I found that TRT outputs is very different from the fake Q/DQ model in python, and the fake Q/DQ onnx model as well when running with onnx runtime. (np.allclose with 1e-3 as the threshold failed)

Are you seeing large divergence between torch model output and onnx runtime output? What is the accuracy degradation  you get from the ONNX model?

**Comment by YixuanSeanZhou** [comment]
Thanks both for the response. 

@realAsma when running with onnx runtime, it only diverges for 0.1% of classification result. But if i am not mistaken, onnx runtime doesn't actually resolve the QDQ nodes into a quantized operation? But i guess it did prove the QDQ in the onnx is, so the bug is not introduced when exporting from torch to onnx 

@riyadshairi979
Following up on yday's discussion, not quantize the first conv in the BasicBlock didn't help. I still see roughly 2% miss classification at the end. I will file a bug report to TRT.

**Comment by i-riyad** [comment]
Closing the issue here since [the same issue](https://github.com/NVIDIA/TensorRT/issues/4079) is resolved with TensorRT.
